src/Sylius/Bundle/AddressingBundle/Repository/ZoneRepository.php

```diff
@@ -0,0 +1,126 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Sylius Sp. z o.o.
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\AddressingBundle\Repository;
+
+use Doctrine\ORM\QueryBuilder;
+use Sylius\Bundle\ResourceBundle\Doctrine\ORM\EntityRepository;
+use Sylius\Component\Addressing\Model\AddressInterface;
+use Sylius\Component\Addressing\Model\Scope;
+use Sylius\Component\Addressing\Model\ZoneInterface;
+use Sylius\Component\Addressing\Repository\ZoneRepositoryInterface;
+
+/**
+ * @implements ZoneRepositoryInterface<ZoneInterface>
+ */
+class ZoneRepository extends EntityRepository implements ZoneRepositoryInterface
+{
+    public function findOneByAddress(AddressInterface $address, ?string $scope = null): ?ZoneInterface
+    {
+        $query = $this->createFindByAddressQueryBuilder($address, $scope);
+
+        $query
+            ->addSelect('(CASE
+                    WHEN z.type = \'province\' THEN 1
+                    WHEN z.type = \'country\' THEN 2
+                    WHEN z.type = \'zone\' THEN 3
+                    ELSE 4
+                END) AS HIDDEN sort_order')
+            ->orderBy('sort_order', 'ASC')
+            ->setMaxResults(1)
+        ;
+
+        return $query->getQuery()->getOneOrNullResult();
+    }
+
+    /** @return ZoneInterface[] */
+    public function findAllByAddress(AddressInterface $address, ?string $scope = null): array
```

Comment:
`findAll` and `findBy` is a bit of contradiction  :slightly_smiling_face:

Generally speaking `findBy` is expected to return collection unless `fineOneBy` is specified, so clarification is not needed.

```patch
--- findAllByAddress
+++ findByAddress
```